### PR TITLE
Deprecate setting handle values with __setattr__ syntax

### LINKED ISSUE #2490: Deprecate __setattr__ overload for setting a handle value
We already have two other ways to accomplish this and the [Zen of Python](https://www.python.org/dev/peps/pep-0020/) states "There should be one-- and preferably only one --obvious way to do it."

```python
handle.value = value
handle <= value
```

The `__setattr__` syntax depends on there being at least one additional level of hierarchy to be able to set a signal; and this can result in unintended behavior with newbies to Python. I have seen this happen at least 3 times.

```python
dut.handle = value  # okay
handle = value  # doh!
```
 
I think this should behavior should be deprecated.

### PR DESCRIPTION
**Pull request by ktbarrett**
Closes #2490. Adds deprecation for `dut.handle = value` syntax.

### CODE REVIEW

**Comment by ktbarrett** [comment]
Must also deprecate `__setitem__` on arrayed objects. Need to find out why the remaining code in `__setattr__` is not being hit.

File: cocotb/handle.py [hunk 1]
```diff
@@ -316,6 +316,10 @@ def __setattr__(self, name, value):
         sub = self.__get_sub_handle_by_name(name)
         if sub is not None:
             sub.value = value
+            warnings.warn(
+                "Setting values on handles using the ``dut.handle = value`` syntax is deprecated."
+                "Instead use the ``handle.value = value`` or ``handle <= value`` syntax",
+                DeprecationWarning, stacklevel=2)
```
Comment (eric-wieser) [inline cocotb/handle.py:322]: I'd swap the order of these lines, so that if the user promotes warnings to errors, the assignment never happens.

**Review by cmarqu — APPROVED** [review]

File: cocotb/handle.py [hunk 2]
```diff
@@ -582,6 +586,10 @@ def __init__(self, handle, path):
     def __setitem__(self, index, value):
         """Provide transparent assignment to indexed array handles."""
         self[index].value = value
+        warnings.warn(
+            "Setting values on handles using the ``dut.handle[i] = value`` syntax is deprecated."
+            "Instead use the ``handle[i].value = value`` or ``handle[i] <= value`` syntax",
+            DeprecationWarning, stacklevel=2)
```
Comment (eric-wieser) [inline cocotb/handle.py:592]: This should happen before the write occurs, not after.

File: documentation/source/newsfragments/2490.removal.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Setting values on handles using the ``dut.handle = value`` syntax is deprecated.
+Instead use the ``handle.value = value`` or ``handle <= value`` syntax.
```
Comment (eric-wieser) [inline documentation/source/newsfragments/2490.removal.rst:2]: Can you do a follow-up to mention `handle[i] = val` syntax being deprecated too?

File: cocotb/handle.py (hunk 2 above)
Comment (ktbarrett) [reply cocotb/handle.py:592]: Damn, I thought I fixed that...

File: documentation/source/newsfragments/2490.removal.rst (hunk 3 above)
Comment (ktbarrett) [reply documentation/source/newsfragments/2490.removal.rst:2]: Sure.

File: cocotb/handle.py (hunk 2 above)
Comment (eric-wieser) [reply cocotb/handle.py:592]: You did, but only for attr not index

Comment (ktbarrett) [reply cocotb/handle.py:592]: Resolved in #2509.

**Pull request merged**